lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp

```diff
@@ -1319,152 +1268,105 @@ void LowerTypesPass::runOnOperation() {
+      // modules and do not target something inside the module.)
+      if (oldRef != path.namepath().getValue().back())
         continue;
-      auto *pathOp = iter->op;
-      bool isInstanceOp = isa<InstanceOp>(pathOp);
-      // There can be multiple annotations that share the same NLA, this list is
-      // used to record all such annotations with the updated NLA reference,
-      // which will be appended to the op annotations.
-      SmallVector<Annotation> newNLAs;
-      auto isModule = isa<FModuleOp>(pathOp) || isa<FExtModuleOp>(pathOp);
-      // If its a module, then get the corresponding port annotation.
-      // This lambda will not be used to update NLAs on modules, because
-      // LowerTypes shouldnot cause any update to them.
-      AnnotationSet annos = isModule
-                                ? AnnotationSet::forPort(pathOp, iter->portIdx)
-                                : AnnotationSet(pathOp);
-      annos.removeAnnotations([&](Annotation anno) {
-        if (auto nlaRef = anno.getMember<FlatSymbolRefAttr>("circt.nonlocal"))
-          if (oldNLAname == nlaRef.getAttr()) {
-            // Found the NLA, now delete it if removeNLA is true.
-            // Else, record its contents with the updated NLA reference.
-            if (removeNLA)
-              return true;
-            auto newNLA = anno;
-            newNLA.setMember("circt.nonlocal", newName);
-            newNLAs.push_back(newNLA);
-            // For InstanceOp, append the reference to the new NLA (i.e., donot
-            // delete). For leaf, just replace with a reference to the new NLA
-            // (i.e., delete old anno).
-            return !isInstanceOp;
-          }
-        return false;
-      });
-      annos.addAnnotations(newNLAs);
-      if (isModule)
-        annos.applyToPort(pathOp, iter->portIdx);
-      else
-        annos.applyToOperation(pathOp);
+
+      // Split the old hierarchical path into one hierarchical path for each new
+      // InnerRefAttr.  Update the symbols in any NLAs which use the old
+      // InnerRefAttr to the correct new InnerRefAttr.
+      auto namepath = path.namepath().getValue();
+      // Grab the old namepath.  We reuse all but the last element of this.
+      SmallVector<Attribute> newNamepath{namepath.begin(), namepath.end()};
+      ImplicitLocOpBuilder builder(path.getLoc(), path);
+      builder.setInsertionPointAfter(path);
+      StringAttr newSym;
+      assert(!newRefs.empty() && "LowerTypes should not delete InnerRefAttrs");
+      for (auto &target : newRefs) {
+        // Drop the last part of the namepath so we can replace it.
+        newNamepath.pop_back();
+
+        // Re-use the old hierarchical path symbol for the first new
+        // hierarchical path.  Generate a new symbol for any later paths.
+        if (!newSym) {
+          newSym = path.getNameAttr();
+          // Delete the old hierarchical path from the NLA and symbol tables.
+          nlaTable->erase(path);
```

Comment:
It might be better if the update to the NLATable is done outside the loop. We can store a list of the `HierPathOp`s that need to be deleted and remove them outside the loop. This might be the cause of the crash you are seeing.

Reply:
Maybe your current implementation is the correct approach. You are already copying the result of the `lookup` on line 1288.
There are issues  in updating the `NLATable` while iterating over the result of a query `lookup`. Because, it returns a reference to the entry. 
If your crash is still present, try to valgrind it !